docker/indexer/stages/preparation/preparation.go

```diff
@@ -0,0 +1,294 @@
+// Package preparation provides functionality to extract tags, branches and commits from repository configurations.
+package preparation
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/google/osv.dev/docker/indexer/shared"
+
+	log "github.com/golang/glog"
+	pb "github.com/google/osv.dev/docker/indexer/proto"
+)
+
+const workers = 5
+
+// Result is the data structure returned by the stage.
+type Result struct {
+	Name            string
+	BaseCPE         string
+	VersionRE       *regexp.Regexp
+	TagMatching     bool
+	BranchMatching  bool
+	CheckoutOptions *git.CheckoutOptions
+	Commit          plumbing.Hash
+	When            time.Time
+	Type            pb.RepositoryType
+	FileExts        []string
+}
+
+// Checker interface is used to check whether a name/hash pair already exists in storage.
+type Checker interface {
+	Exists(ctx context.Context, name string, hash plumbing.Hash) (bool, error)
+}
+
+// Stage holds the data types necessary to process repository configuration.
+type Stage struct {
+	Checker Checker
+	RepoHdl *storage.BucketHandle
+}
+
+// Run runs the stage and outputs Result data types to the results channel. 
+func (s *Stage) Run(ctx context.Context, cfgs []*pb.Repository, results chan *Result) error {
+	var err error
+	wErr := make(chan error, workers)
+	wg := sync.WaitGroup{}
+	routineCtr := 0
+	wCtx, wCancel := context.WithCancel(ctx)
+	for _, repoCfg := range cfgs {
+		if routineCtr >= workers {
+			wg.Wait()
```

Comment:
Can we just use a semaphore here (i.e. https://pkg.go.dev/golang.org/x/sync/semaphore) ?